x-pack/plugins/apm/server/lib/search_strategies/correlations/async_search_service.ts

```diff
@@ -0,0 +1,265 @@
+  histogram: HistogramItem[];
+  value: string;
+  field: string;
+  correlation: number;
+  ksTest: number;
+  duplicatedFields: Set<string>;
+}
+export const asyncSearchServiceProvider = (
+  esClient: ElasticsearchClient,
+  params: SearchServiceParams
+) => {
+  let isCancelled = false;
+  let isRunning = true;
+  let error: Error;
+
+  const progress: AsyncSearchProviderProgress = {
+    started: Date.now(),
+    loadedHistogramStepsize: 0,
+    loadedOverallHistogram: 0,
+    loadedFieldCanditates: 0,
+    loadedFieldValuePairs: 0,
+    loadedHistograms: 0,
+    getOverallProgress: () =>
+      progress.loadedHistogramStepsize * 0.025 +
+      progress.loadedOverallHistogram * 0.025 +
+      progress.loadedFieldCanditates * 0.025 +
+      progress.loadedFieldValuePairs * 0.025 +
+      progress.loadedHistograms * 0.9,
+  };
+
+  const values: SearchServiceValue[] = [];
+  let overallHistogram: HistogramItem[] | undefined;
+
+  let percentileThresholdValue: number;
+
+  const cancel = () => {
+    isCancelled = true;
+  };
+
+  const fetchCorrelations = async () => {
+    try {
+      // 95th percentile to be displayed as a marker in the log log chart
+      const percentileThreshold = await fetchTransactionDurationPecentiles(
+        esClient,
+        params,
+        params.percentileThreshold ? [params.percentileThreshold] : undefined
+      );
+      percentileThresholdValue =
+        percentileThreshold[`${params.percentileThreshold}.0`];
+
+      const histogramRangeSteps = await fetchTransactionDurationHistogramRangesteps(
+        esClient,
+        params
+      );
+      progress.loadedHistogramStepsize = 1;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const overallLogHistogramChartData = await fetchTransactionDurationRanges(
+        esClient,
+        params,
+        histogramRangeSteps
+      );
+      progress.loadedOverallHistogram = 1;
+      overallHistogram = overallLogHistogramChartData;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const {
+        fieldCandidates,
+        totalHits,
+      } = await fetchTransactionDurationFieldCandidates(esClient, params);
+
+      progress.loadedFieldCanditates = 1;
+
+      // Create an array of ranges [2, 4, 6, ..., 98]
+      const percents = Array.from(range(2, 100, 2));
+      const percentilesRecords = await fetchTransactionDurationPecentiles(
+        esClient,
+        params,
+        percents
+      );
+      const percentiles = Object.values(percentilesRecords);
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const fieldValuePairs = await fetchTransactionDurationFieldValuePairs(
+        esClient,
+        params,
+        fieldCandidates,
+        progress,
+        percentiles
+      );
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      async function* fetchTransactionDurationHistograms() {
```

Comment:
Any reason to use generators instead of async/await ? (asking because I'm not very familiar with generators)

Reply:
Not a special reason, also a first for us, I was interesting in trying out `for await`, I thought the readability of `for await (const item of fetchTransactionDurationHistograms()) { ... }` turned out quite nice. `fetchTransactionDurationHistograms()` still lives inline here but we might want to move it to a separate file in a follow up (it's more an implementation detail) and which should make this file a little be more concise. Just let us know if you're fine to keep it this way.